src/org/jgroups/raft/filelog/FilePositionCache.java

```diff
@@ -11,77 +12,222 @@
-   public static final int OK = 1;
-   private static final int SIZE_INCREMENT = 128;
 
-   private final long[] position;
+   private static final int PAGE_CAPACITY = 1 << 10;
+   private static final int PAGE_SHIFT = Integer.numberOfTrailingZeros(PAGE_CAPACITY);
+   private static final int PAGE_MASK = PAGE_CAPACITY - 1;
+   private static final int EMPTY = -1;
+
+   {
+      assert Integer.bitCount(PAGE_CAPACITY) == 1 : "PAGE_CAPACITY MUST BE A POWER OF TWO";
+   }
+
+   private static final class PositionPage {
+
+      private final long[] positions;
+      private int lastNotEmptyOffset;
+
+      public PositionPage() {
+         positions = new long[PAGE_CAPACITY];
+         Arrays.fill(positions, EMPTY);
+         lastNotEmptyOffset = -1;
+      }
+
+      public boolean clear() {
+         return clearFrom(0);
+      }
+
+      public boolean clearFrom(final int pageOffset) {
+         if (pageOffset > lastNotEmptyOffset) {
+            return false;
+         }
+         if (pageOffset < 0) {
+            throw new IllegalArgumentException("pageOffset must be greater then 0");
+         }
+         if (pageOffset >= PAGE_CAPACITY) {
+            throw new IllegalArgumentException("pageOffset must be less then page capacity");
+         }
+         assert lastNotEmptyOffset >= 0;
+         Arrays.fill(positions, pageOffset, lastNotEmptyOffset + 1, EMPTY);
+         lastNotEmptyOffset = -1;
+         // must search backward for the first not empty slot, to mark it
+         for (int i = pageOffset; i >= 0; i--) {
+            if (positions[i] != EMPTY) {
+               lastNotEmptyOffset = i;
+               break;
+            }
+         }
+         return true;
+      }
+
+      public long set(final int pageOffset, final long position) {
+         if (position < 0) {
+            throw new IllegalArgumentException("position must be greater then zero");
+         }
+         if (pageOffset >= PAGE_CAPACITY) {
+            throw new IllegalArgumentException("The required pageOffset is beyond page capacity");
+         }
+         final long oldValue = positions[pageOffset];
+         if (oldValue == EMPTY) {
+            if (pageOffset > lastNotEmptyOffset) {
+               lastNotEmptyOffset = pageOffset;
+            }
+         }
+         positions[pageOffset] = position;
+         return oldValue;
+      }
+
+      public long get(final int pageOffset) {
+         if (pageOffset >= PAGE_CAPACITY) {
+            throw new IllegalArgumentException("The required pageOffset is beyond page capacity");
+         }
+         if (pageOffset > lastNotEmptyOffset) {
+            return EMPTY;
+         }
+         return positions[pageOffset];
+      }
+   }
+
+   private final ArrayList<PositionPage> positionPages;
    private final int firstLogIndex;
 
    public FilePositionCache(int firstLogIndex) {
-      this.position = new long[SIZE_INCREMENT];
+      this.positionPages = new ArrayList<>();
```

Comment:
can we "allocate" a couple of pages? `ArrayList` starts with an empty array.

Reply:
Yep, I've made it on purpose for the sparse entry use case (that AFAIK is a bad use case, and shouldn't happen really): pages are materialized only if any of its element need has been set